### ISSUE TITLE
How do you refresh a token?

### ISSUE THREAD

**Issue by alessandroderomawep** [HyperSense-Software/hss-okta-flutter#38]
I'm getting the token with this code on web:

```
oktaWeb.auth.token.getWithRedirect(
  AuthorizeOptions(
    responseType: ['token', 'id_token'],
    scopes: ['openid', 'profile', 'email', 'offline_access'],
    nonce: 'my generated nonce',
  ),
),
```

Using this to store tokens after successful authentication:

```
oktaWeb.auth.storeTokensFromRedirect();
```

Everything works, except the token expires after 1h. The sdk doesn't handle auto-renewal because everytime I call `getIdToken()` it returns the expired token. I see the session is still active because if I redirect the user to the okta authentication page it immediately passes authentication without even asking for credentials, meaning the session is still valid even if the token expired. 

I need to silently refresh the idToken, but the `renewToken` funciton seems to expect a key, which I don't know what it is and where to find it.

How do I renew the token?

**Comment by aldrin233** [comment]
@alessandroderomawep thanks for your report, it seems that I made a mistake with the parameters of renewToken, it needs an implementation of AbstractToken as its parameters. 

what you can do is turn you current token into a json and use it with renewToken

**Comment by aldrin233** [comment]
i will issue an update correcting this mistake.

**Comment by alessandroderomawep** [comment]
@aldrin233 thank you for your support. Could you by any chance post an example code of the flow you just described?

**Comment by aldrin233** [comment]
```
oktaWeb.auth.token.getWithRedirect(
  AuthorizeOptions(
    responseType: ['token', 'id_token'],
    scopes: ['openid', 'profile', 'email', 'offline_access'],
    nonce: 'my generated nonce',
  ),
),

final token = /// get your token somehow
final tokenJson = jsonEncode(token)
final newToken = await oktaWeb.auth.token.renew(tokenJson)
```

**Comment by alessandroderomawep** [comment]
@aldrin233 thank you. What is `token` here? The full token response? Just the idToken/accessToken?

**Comment by aldrin233** [comment]
@alessandroderomawep it's this
```
// this token is provided by Okta via getWithoutPrompt, getWithPopup, and parseFromUrl

var tokenToRenew = {
  idToken: 'YOUR_ID_TOKEN_JWT',
  claims: { /* token claims */ },
  expiresAt: 1449699930,
  scopes: ['openid', 'email'],
  authorizeUrl: 'https://{yourOktaDomain}/oauth2/v1/authorize',
  issuer: 'https://{yourOktaDomain}',
  clientId: 'NPSfOkH5eZrTy8PMDlvx'
};
```

**Comment by alessandroderomawep** [comment]
@aldrin233 we tried with your method:

```
Future<TokenResponse> refreshToken() async {
    final IDToken? token = (await oktaWeb.getTokens()).idToken;
    final String tokenRefresh = jsonEncode({
      "idToken": token?.idToken,
      "claims": {
        "authTtime": token?.claims.authTtime,
        "aud": token?.claims.aud,
        "email": token?.claims.email,
        "emailVerified": token?.claims.emailVerified,
        "exp": token?.claims.exp,
        "familyName": token?.claims.familyName,
        "givenName": token?.claims.givenName,
        "iat": token?.claims.iat,
        "iss": token?.claims.iss,
        "jti": token?.claims.jti,
        "locale": token?.claims.locale,
        "name": token?.claims.name,
        "nonce": token?.claims.nonce,
        "preferredUsername": token?.claims.preferredUsername,
        "sub": token?.claims.sub,
        "updatedAt": token?.claims.updatedAt,
        "ver": token?.claims.ver,
        "zoneinfo": token?.claims.zoneinfo,
        "atHash": token?.claims.atHash,
        "acr": token?.claims.acr,
      },
      "expiresAt": token?.expiresAt,
      "scopes": token?.scopes,
      "authorizeUrl": token?.authorizeUrl,
      "issuer": token?.issuer,
      "clientId": token?.clientId
    });
    print(tokenRefresh); // prints correct token with same structure as yours
    return await promiseToFuture<TokenResponse>(oktaWeb.auth.token.renew(tokenRefresh));
}
```

Unfortunately we get this error:

```
Error: AuthSdkError: Renew must be passed a token with an array of scopes and an accessToken or idToken
```

Any clues?

**Comment by alessandroderomawep** [comment]
@aldrin233 If you look at my code I inserted a comment here:

```
print(tokenRefresh); // prints correct token with same structure as yours
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I'm removing sensitive informations, but our object after jsonEncode looks like this:

```
{
  "idToken": "eyJraWQiOiJJT20x<redacted>5w",
  "claims": {
    "authTtime": null,
    "aud": "<redacted>",
    "email": "<redacted>",
    "emailVerified": null,
    "exp": 1712834137,
    "familyName": null,
    "givenName": null,
    "iat": 1712830537,
    "iss": "https://<redacted>",
    "jti": "<redacted>",
    "locale": null,
    "name": "<redacted>",
    "nonce": "<redacted>",
    "preferredUsername": null,
    "sub": "<redacted>",
    "updatedAt": null,
    "ver": 1,
    "zoneinfo": null,
    "atHash": null,
    "acr": null
  },
  "expiresAt": 1712834116,
  "scopes": ["openid", "profile", "email", "offline_access"],
  "authorizeUrl": "https://<redacted>/oauth2/v1/authorize",
  "issuer": "https://<redacted>",
  "clientId": "<redacted>"
}
```

The object looks good to me as per your previous comment.
Any clues? Am I not seeing something obvious?

**Comment by aldrin233** [comment]
hmm, if you are passing the correct structure there is not much more we can do but wait for my update.

**Comment by aldrin233** [comment]
Issue resolved on version 1.4.0
